### PR TITLE
Identity basis

### PR DESCRIPTION
**Pull request by BalzaniEdoardo**
In this PR I add an identity basis function and a basis function that capture raw history.

I added a few how-to guide that showcase the use of these basis for:

1. Including custom predictors in a composite basis.
2. Create a design from categorical variables.
3. Fit a coupled GLM.

### CODE REVIEW

File: src/nemos/basis/_identity.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        vmax = np.nanmax(sample_pts, axis=0) if self.bounds is None else self.bounds[1]
+        sample_pts[(sample_pts < vmin) | (sample_pts > vmax)] = np.nan
+        check_fraction_valid_samples(
+            sample_pts,
+            err_msg="All the samples lie outside the [vmin, vmax] range.",
+            warn_msg="More than 90% of the samples lie outside the [vmin, vmax] range.",
+        )
+        return sample_pts[..., np.newaxis]
+
+    def evaluate_on_grid(self, n_samples: int) -> Tuple[Tuple[NDArray], NDArray]:
```
Comment (billbrod) [inline src/nemos/basis/_identity.py:72]: I think return type is wrong? shouldn't it just be `Tuple[NDarray, NDArray]`?

File: src/nemos/basis/_identity.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            The equi-spaced sample points X as a 2D array (n_samples, 1).
+        """
+        return super().evaluate_on_grid(n_samples)
+
+    def _check_n_basis_min(self):
+        """No checks necessary."""
+        pass
+
+    @property
+    def n_basis_funcs(self) -> tuple | None:
```
Comment (billbrod) [inline src/nemos/basis/_identity.py:96]: why is this necessary? because it's read-only?

File: src/nemos/basis/_identity.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        """Read-only property for identity."""
```
Comment (billbrod) [inline src/nemos/basis/_identity.py:97]:
```suggestion
        """Number of basis functions (read-only)."""
```

File: src/nemos/basis/basis.py [hunk 4]
```diff
@@ -1830,3 +1825,248 @@ def _check_window_size(self, window_size: int):
...
+        >>> _ = basis.set_input_shape(x)
+        >>> basis.n_output_features
+        20
+
+        """
+        return AtomicBasisMixin.set_input_shape(self, xi)
+
+
+class HistoryConv(ConvBasisMixin, HistoryBasis):
+    """Identity basis function.
```
Comment (billbrod) [inline src/nemos/basis/basis.py:1947]: Need a better one-liner here, at least to distinguish with IdentityEval

File: docs/background/basis/README.md [hunk 5]
```diff
@@ -64,6 +64,21 @@
      - :class:`~nemos.basis.OrthExponentialEval`  :raw-html:`<br />`
        :class:`~nemos.basis.OrthExponentialConv`
      - 🟢 Eval
+   * - **Identity Function**
```
Comment (billbrod) [inline docs/background/basis/README.md:67]: Feels odd to have these two be separate, as they're then the only ones that don't have both modes. Why not include them in the same row?

File: docs/how_to_guide/categorical_predictors.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
+# Construct Design Matrices for Categorical Features
+
+## Design Matrix with `patsy`
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:16]: should be clear here that `patsy` is optional, like scikit learn, but that might be hlpeful

File: docs/how_to_guide/categorical_predictors.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+In the design matrix:
+- `T.` indicates **treatment coding**, which is used to encode categorical variables. For example, `stimulus[T.s2]` represents the presence (`1`) or absence (`0`) of the category `s2` compared to the reference category `s1`.
+- The reference category (`s1`) is dropped to avoid collinearity, as including all categories would result in redundancy.
+- Similarly, for `context`, columns `context[c1]` and `T.context[c2]` represent the presence of each category. 
+- The interaction term `stimulus[T.s2]:context[T.c2]` represents the combined effect of `stimulus = s2` and `context = c2`. Only one column is needed for the interaction, as the other combinations are implicitly represented by the reference categories (`s1` and `c1`).
+:::
+- 
+NeMoS GLMs, however, already specify an intercept term, making the first column of this dataframe redundant. Let's drop it.
+
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:69]: this is rendered weirdly -- what's the `-` doing here? I also think this should be part of hte paragraph befor ethe admonition.

I think the admonition should be moved to right after the code box, the title needs to be corrected (it's currently not rendering as the title), and the admonition should probably have a link to the relevant part of the patsy docs: "See patsy docs for more details"

File: docs/how_to_guide/categorical_predictors.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+The feature in the design matrix are 1-hot encoded categories.
+Note that `patsy` adds an intercept and drops the reference `s1`. This is done by design to avoid perfect collinearity (the sum of the `s1` and the `s2` column would be equal to the intercept). 
+
+:::{note} Understanding `patsy`'s Output
+:class: dropdown
+
+In the design matrix:
+- `T.` indicates **treatment coding**, which is used to encode categorical variables. For example, `stimulus[T.s2]` represents the presence (`1`) or absence (`0`) of the category `s2` compared to the reference category `s1`.
+- The reference category (`s1`) is dropped to avoid collinearity, as including all categories would result in redundancy.
+- Similarly, for `context`, columns `context[c1]` and `T.context[c2]` represent the presence of each category.
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:64]:
```suggestion
- Similarly, for `context`, columns `context[c1]` and `T.context[c2]` represent the presence of each category, and `context[T.c2]` is dropped to avoid collinearity. 
```

File: docs/how_to_guide/categorical_predictors.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+Let's assume we have a simple dataset with four samples and two labels assigned to each sample. You can think of these labels as characteristics of the trial to which the samples belong, such as stimulus 1 vs stimulus 2 or context 1 vs context 2.
+
+```{code-cell} ipython3
+# Example data
+import pandas as pd
+from patsy import dmatrix
+import numpy as np
+
+# Example data 
+data = pd.DataFrame({
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:29]: this example is very abstract. I think we'd better off using specific stimulus / context, and pointing out that the general principle applies

File: docs/how_to_guide/categorical_predictors.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+:class: dropdown
+
+In the design matrix:
+- `T.` indicates **treatment coding**, which is used to encode categorical variables. For example, `stimulus[T.s2]` represents the presence (`1`) or absence (`0`) of the category `s2` compared to the reference category `s1`.
+- The reference category (`s1`) is dropped to avoid collinearity, as including all categories would result in redundancy.
+- Similarly, for `context`, columns `context[c1]` and `T.context[c2]` represent the presence of each category. 
+- The interaction term `stimulus[T.s2]:context[T.c2]` represents the combined effect of `stimulus = s2` and `context = c2`. Only one column is needed for the interaction, as the other combinations are implicitly represented by the reference categories (`s1` and `c1`).
+:::
+- 
+NeMoS GLMs, however, already specify an intercept term, making the first column of this dataframe redundant. Let's drop it.
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:68]:
```suggestion
NeMoS GLMs, however, implicitly specify an intercept term, making the first column of this dataframe redundant. Let's drop it.
```

File: docs/how_to_guide/categorical_predictors.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+design_df.drop(columns=["Intercept"], inplace=True)
+
+design_df
+```
+
+
+
+
+### Fit the GLM
+
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:80]: need some text here explaining what's happening

File: docs/how_to_guide/categorical_predictors.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+```{code-cell} ipython3
+import nemos as nmo
+
+# Fit the GLM model
+model = nmo.glm.GLM().fit(design_df, counts)
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:85]: this is a bit confusing, because we don't talk about pandas dataframes anywhere else. I'm assuming it works because we just grab `design_df.values`, the array representation, but let's make that explicit.

File: docs/how_to_guide/categorical_predictors.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+## Categorical Predictors and Basis Functions
+
+Assume you have an additional time series that you want to process with a basis:
+
+```{code-cell} ipython3
+# Example time series
+speed = np.array([10., 3., 2., 20.])
+```
+
+### Composite Basis for Multiple Predictors
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:97]: this header doesn't add anything

File: docs/how_to_guide/categorical_predictors.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+You can create a composite basis to combine predictors:
+
+```{code-cell} ipython3
+# Identity basis combined with a B-spline basis
+bas = nmo.basis.IdentityEval() + nmo.basis.BSplineEval(5)
+
+# Compute features
+X = bas.compute_features(design_df, speed)
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:106]: should describe what this does: identity basis just passes through the design_df (again, I think we should add `.values`), while `BSplineEval` works with speed.

and probably link to the bit about additive bases

maybe show the array? or plot it?

File: docs/how_to_guide/categorical_predictors.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+```
+
+### Interaction with Basis algebra
+
+To model the interaction between two variables, you can take advantage of the basis multiplication:
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:111]: this is an additional interaction right? so maybe say something like 
```suggestion
To add the interaction between the `c2` context and speed to your design matrix, you can take advantage of basis multiplication:
```

File: docs/how_to_guide/categorical_predictors.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+```{code-cell} ipython3
+# Interaction basis
+bas = nmo.basis.IdentityEval() * nmo.basis.BSplineEval(5)
+
+# Compute features for interaction
+X2 = bas.compute_features(design_df["context[T.c2]"], speed)
+```
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:119]: and again, I think we should visualize this output

File: docs/how_to_guide/custom_predictors.md [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
+(custom-features)=
+# Custom Features
```
Comment (billbrod) [inline docs/how_to_guide/custom_predictors.md:15]: I don't think "custom" is the right word here, but I don't have a better one.

File: docs/how_to_guide/custom_predictors.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+## Incorporating Custom Features into GLM Design
+
+In some cases, your data may contain features that are not directly computable through the available bases in NeMoS, for instance a Principal Component Analysis (PCA) of high-dimensional signals. 
+The `IdentityEval` basis allows you to incorporate these precomputed features into the model design, and combine them with other predictors through basis composition.
+
+For example, let's assume that you want to compute a Principal Component Analysis (PCA) of some signals, and use 
+the first 2 Principal Components (PCs) as GLM features.
```
Comment (billbrod) [inline docs/how_to_guide/custom_predictors.md:23]:
```suggestion
the first 2 Principal Components as GLM features. 
```

File: docs/how_to_guide/custom_predictors.md [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+## Incorporating Custom Features into GLM Design
+
+In some cases, your data may contain features that are not directly computable through the available bases in NeMoS, for instance a Principal Component Analysis (PCA) of high-dimensional signals. 
+The `IdentityEval` basis allows you to incorporate these precomputed features into the model design, and combine them with other predictors through basis composition.
+
+For example, let's assume that you want to compute a Principal Component Analysis (PCA) of some signals, and use 
+the first 2 Principal Components (PCs) as GLM features. 
+
+Currently, NeMoS doesn't provide a PCA basis, but you can compute the PCs in `sklearn`.
```
Comment (billbrod) [inline docs/how_to_guide/custom_predictors.md:25]:
```suggestion
Currently, NeMoS doesn't provide a PCA basis, but you can compute the principle components in `sklearn`.
```

File: docs/how_to_guide/custom_predictors.md [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+# generate some counts
+counts = np.random.poisson(size=n_samples)
+
+# compute the first 2 pcs
+pcs = PCA(2).fit_transform(high_dim_singals)
+
+```
+
+Now, let's see how to use the `IdentityEval` basis to model jointly the PCs and a spike history filter.
```
Comment (billbrod) [inline docs/how_to_guide/custom_predictors.md:45]:
```suggestion
Now, let's see how to use the `IdentityEval` basis to model jointly the principle components and a spike history filter.
```

File: docs/how_to_guide/custom_predictors.md [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+```{code-cell}
+import nemos as nmo
+
+# create a composite basis 
+pc_basis = nmo.basis.IdentityEval()
+history_basis = nmo.basis.RaisedCosineLogConv(3, window_size=10)
+composite_basis = pc_basis + history_basis
+
+# create the model design
+X = composite_basis.compute_features(pcs, counts)
```
Comment (billbrod) [inline docs/how_to_guide/custom_predictors.md:56]: I feel like we should visualize this, or at least show the shape?

File: docs/how_to_guide/raw_history_feature.md [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+
+plt.title("Basis Feature Matrix", fontsize=16)
+plt.imshow(X2[10:], cmap="Greys")
+plt.xticks([1, 4, 7, 10], [0, 1, 2, 3])
+plt.xlabel("Neurons", fontsize=12)
+plt.ylabel("Samples", fontsize=12)
+plt.show()
+```
+
+This model can be fit with the usual `model_reduced = nmo.glm.GLM().fit(X2, counts)`.
```
Comment (billbrod) [inline docs/how_to_guide/raw_history_feature.md:93]: should be population glm, right?

File: docs/how_to_guide/raw_history_feature.md [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+mask[1, 0] = 0
+
+# remove the directional coupling from 2 to 3
+mask[2, 3] = 0
+
+# repeat over the rows by the number of basis functions
+mask = np.repeat(mask, basis.n_basis_funcs, axis=0)
+```
+
+We are now readi to fit the GLM masking the coefficients.
```
Comment (billbrod) [inline docs/how_to_guide/raw_history_feature.md:142]:
```suggestion
We are now ready to fit the GLM masking the coefficients.
```

File: docs/how_to_guide/raw_history_feature.md
Comment (billbrod) [inline docs/how_to_guide/raw_history_feature.md:1]: why not do this one using the data from https://nemos--282.org.readthedocs.build/en/282/tutorials/plot_02_head_direction.html? then we can visualize the outputs (and point them to the full tutorial for more details)

Comment (billbrod) [reply docs/how_to_guide/raw_history_feature.md:1]: regardless, I think we should visualize the model coefficients we get

**Review by billbrod — CHANGES_REQUESTED** [review]
It looks like this PR is showing the changes from #280 in the diff, do you know why?

Otherwise, these looks generally good, though I think this means we're going to have to go through the existing how-tos ,etc to see where to use the new bases (at least in the head direction tutorial). Can we make an issue for that?

The new how-tos all looks like they match what they are in diataxis, i.e., very short and answering a single question. I think we'll need to go back through the existing ones and update / move them then, and I think we should consider changing the layout of the landing page for them: make the cards smaller, only show the top-level title, which should be very descriptive. This should also be a separate issue.

File: src/nemos/basis/_identity.py (hunk 1 above)
Comment (BalzaniEdoardo) [reply src/nemos/basis/_identity.py:72]: Yes!

File: src/nemos/basis/_identity.py (hunk 2 above)
Comment (BalzaniEdoardo) [reply src/nemos/basis/_identity.py:96]: yes, to make it read-only

File: src/nemos/basis/basis.py (hunk 4 above)
Comment (BalzaniEdoardo) [reply src/nemos/basis/basis.py:1947]: yep, copy-paste issue

File: docs/background/basis/README.md (hunk 5 above)
Comment (BalzaniEdoardo) [reply docs/background/basis/README.md:67]: My only point here is that history is like a lagged version of the identity, so it is not exactly the "conv" version of identity. 

If it was that, i would expect that the kernel  for convolution would be just a single `np.linsapce(0,1,n_samples)` (which is f(x) = x when x is the linspace).

Instead the history basis is a bunch of delta function that peaks at different lags. 

Also the image in the table of identity doesn't apply to the history so we would not be able to have an image that makes sense for both.

What do you think?

File: docs/how_to_guide/categorical_predictors.md (hunk 6 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/categorical_predictors.md:16]: I added a dropdown admonition

File: docs/how_to_guide/categorical_predictors.md (hunk 9 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/categorical_predictors.md:29]: I tried to make the stim and context more concrete, let me know if it works

File: docs/how_to_guide/categorical_predictors.md (hunk 12 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/categorical_predictors.md:85]: added an admonition about it

Comment (BalzaniEdoardo) [reply docs/how_to_guide/categorical_predictors.md:85]: let me know if it works

File: docs/how_to_guide/categorical_predictors.md (hunk 14 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/categorical_predictors.md:106]: the `.values` is not needed, i think it's extra syntax that they should not care about. If the dataframe contains non-numerical values it will fail.

I'll add a description and plot the content

File: docs/how_to_guide/categorical_predictors.md (hunk 16 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/categorical_predictors.md:119]: I wrapped the output into a DataFrame to add back column names and have a more readable output

File: docs/how_to_guide/custom_predictors.md (hunk 17 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/custom_predictors.md:15]: Precomputed?

File: docs/how_to_guide/custom_predictors.md (hunk 21 above)
Comment (BalzaniEdoardo) [reply docs/how_to_guide/custom_predictors.md:56]: I think i simulated only a 1D array actually

File: docs/how_to_guide/raw_history_feature.md
Comment (BalzaniEdoardo) [reply docs/how_to_guide/raw_history_feature.md:1]: i think it requires too much pre-processing  for this session. I'll print the coefficients

Comment (BalzaniEdoardo) [reply docs/how_to_guide/raw_history_feature.md:1]: actually, I take it back... i can just grab the first sec or so and use that data

Comment (BalzaniEdoardo) [reply docs/how_to_guide/raw_history_feature.md:1]: also, i would need more explaining about the data, maybe filter a subset of unit and samples... I think it is better to go straight to the point. If we had a sample data which has already a sub-population and a short duration we could load that (like the iris dataset for sklearn)

File: docs/background/basis/README.md (hunk 5 above)
Comment (billbrod) [reply docs/background/basis/README.md:67]: I think that makes sense.

File: docs/how_to_guide/categorical_predictors.md (hunk 9 above)
Comment (billbrod) [reply docs/how_to_guide/categorical_predictors.md:48]: I think this is good!

File: docs/how_to_guide/categorical_predictors.md (hunk 12 above)
Comment (billbrod) [reply docs/how_to_guide/categorical_predictors.md:109]: added some suggetions

File: docs/how_to_guide/categorical_predictors.md [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+import nemos as nmo
+
+# Fit the GLM model
+model = nmo.glm.GLM().fit(design_df, counts)
+```
+:::{note}
+
+You can directly pass a [`pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) to the 
+`fit` method, as long as it can be internally converted to a floating-point array. The conversion is handled automatically, 
+so no additional steps are required on your part.
```
Comment (billbrod) [inline docs/how_to_guide/categorical_predictors.md:115]:
```suggestion
You can directly pass a [`pandas.DataFrame`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) to the 
`fit` method, as long as it can be internally converted to a floating-point array (that means, for example, that you can't have a column whose values are strings). The conversion is handled automatically, by calling `design_df.values`,
so no additional steps are required on your part.
```

File: docs/how_to_guide/categorical_predictors.md (hunk 14 above)
Comment (billbrod) [reply docs/how_to_guide/categorical_predictors.md:134]: but it will fail if they try to do use a dataframe with non-numeric values anyway, right?

Comment (billbrod) [reply docs/how_to_guide/categorical_predictors.md:134]: I actually don't think we should display it as a dataframe -- we should show what the output looks like: an appropriately shaped array.

File: docs/how_to_guide/categorical_predictors.md (hunk 16 above)
Comment (billbrod) [reply docs/how_to_guide/categorical_predictors.md:162]: same point about not showing it as a dataframe. just show it as an array and explain what the columns mean in words

File: docs/how_to_guide/raw_history_feature.md
Comment (billbrod) [reply docs/how_to_guide/raw_history_feature.md:1]: Okay, I think this is fine, but it would be nice to think about whether there is a meaningful dataset we could use. because it's nice to plot the coefficients and have them look meaningful.

**Review by billbrod — APPROVED** [review]

File: docs/how_to_guide/raw_history_feature.md
Comment (BalzaniEdoardo) [reply docs/how_to_guide/raw_history_feature.md:1]: you are right

**Comment by BalzaniEdoardo** [comment]
> ?



> It looks like this PR is showing the changes from #280 in the diff, do you know why?
No idea, I might have branched from `set_shape_basis_method` and this confused the history. I am not sure it makes sense

> Otherwise, these looks generally good, though I think this means we're going to have to go through the existing how-tos ,etc to see where to use them. Can we make an issue for that?
done

> The new how-tos all looks like they match what they are in diataxis, i.e., very short and answering a single question. I think we'll need to go back through the existing ones and update / move them then, and I think we should consider changing the layout of the landing page for them: make the cards smaller, only show the top-level title, which should be very descriptive. This should also be a separate issue.

I agree with both points
